### DISCUSSION TITLE
Ping measurement for CM and GC servers

### DISCUSSION THREAD

**Discussion by knkitten** [SteamRE/SteamKit#1080] [General]
Аfter successful login I get a message `EMsg.ClientCMList`. This message contains (in my case) 80 CM server addresses (`uint32_t *cm_addresses` and `uint32_t *cm_ports` arrays). I want to choose among all of them the best CM server in terms of minimum latency. What can be used to do this? 

I am also interested in this question but with GC (Dota2). First of all, do we have a choice of GC like CM server? To establish a session with the GC I send first `EMsg.ClientGamesPlayed` with `game_id = 570` and after receiving `EMsg.ClientPlayingSessionState` with `playing_app = 570` I send `EMsg.GCClientHello` and wait `EMsg.GCClientWelcome`. Аfter that, I want the GC response to be as quick as possible. Do I understand correctly that this only depends on the right choice of CM server (with the lowest ping)?

**Comment by yaakov-h** [comment]
If you have a particular CM you want to connect to, you can call `client.Connect(ServerRecord)` to connect to that particular one.

GC sessions are established implicitly via the CM, to my knowledge you cannot choose anything about it. Potentially you could connect to CM in New York and have a Game Coordinator in London if Steam's data centers allowed for and decided such a thing, the client would be none the wiser.

**Reply by knkitten** [reply]
Yes I can connect to a particular CM. However, I want to choose a CM server with minimal delay/ping from a list of known CM servers (from `EMsg.ClientCMList`). Are there any tools to do this?

**Reply by yaakov-h** [reply]
You can write whatever code you want to choose a particular CM server.
